### PR TITLE
Updates on GEM onlineDQM due to the updates of DAQ dataformat

### PR DESCRIPTION
**Pull request by quark2**
#### PR description:
Since lots of stuffs in DAQ dataformat have been changed in https://github.com/cms-sw/cmssw/pull/34504, the GEM onlineDQM has to be changed to adjust it. This PR is for it.

Also many of plots are rearranged, and new plots are added, which are plots of the average cluster sizes, etc..

We also have a plan to make a backport of this update, for the DAQ dataformat (see https://github.com/cms-sw/cmssw/pull/34585), into CMSSW_11_3_X for current cosmic runs.

#### PR validation:
Test are done and one can check again by `runTheMatrix` workflows

#### if this PR is a backport please specify the original PR and why you need to backport that PR:
To CMSSW_11_3_X to be deployed on current cosmic runs

@jshlee @watson-ij @hyunyong

### CODE REVIEW

**Comment by perrotta** [comment]
please test

**Comment by qliphy** [comment]
please test

File: DQM/GEM/interface/GEMDQMBase.h [hunk 1]
```diff
@@ -237,13 +299,16 @@ class GEMDQMBase : public DQMEDAnalyzer {
 
     dqm::impl::MonitorElement *FindHist(K key) {
       if (mapHist.find(key) == mapHist.end()) {
-        std::cout << "" << std::endl;  // FIXME: It's about sending a message
+        std::cout << "WARNING: Cannot find the histogram corresponing to the given key"
```
Comment (jfernan2) [inline DQM/GEM/interface/GEMDQMBase.h:302]: I am sorry but couts are discouraged, please use edm::Log or similar instead

**Comment by jfernan2** [comment]
@quark2 I am confused by this statement:

> We also have a plan to make a backport of this update, for the DAQ dataformat (see #34585), into CMSSW_11_3_X for current cosmic runs.

since #34585 does not contain the DQM/Validation changes from this PR. Do you plan to use this in CRUZET/CRAFT? If so we need a backport

**Comment by jfernan2** [comment]
There are plots with entries but mean = 0, is this expected?

https://tinyurl.com/ydw8hnar  (at the bottom)
https://tinyurl.com/yjstnhvf

**Comment by quark2** [comment]
Hi @jfernan2

> @quark2 I am confused by this statement:
> 
> > We also have a plan to make a backport of this update, for the DAQ dataformat (see #34585), into CMSSW_11_3_X for current cosmic runs.
> 
> since #34585 does not contain the DQM/Validation changes from this PR. Do you plan to use this in CRUZET/CRAFT? If so we need a backport

Some nomenclature rules for GEM are changed, which is shared in GEM onlineDQM/offlineDQM/Validation. And yep, we plan to use it in CRUZET/CRAFT. Should I make a backport PR now?

**Comment by quark2** [comment]
> There are plots with entries but mean = 0, is this expected?
> 
> https://tinyurl.com/ydw8hnar (at the bottom)
> https://tinyurl.com/yjstnhvf

Although I have no idea about the details of the simulation, I think it has no problem since there is no occupancy of both of pads and digis.

**Comment by jfernan2** [comment]
Well, for the moment CRUZET is using 11_3_X so yes but, have you made some validation for the Online GEM clients? I remind you that runTheMatrix runs only Offline DQM, and you are getting ~800 empty plots

**Comment by quark2** [comment]
Yep, the GEM onlineDQM part has been validated by clients. But I have no idea about the empty plots in offlineDQM yet. I'll check it with GEM offlineDQM experts.

File: DQM/GEM/interface/GEMDQMBase.h (hunk 1 above)
Comment (quark2) [reply DQM/GEM/interface/GEMDQMBase.h:302]: Oops, okay. I'll fix it.

**Comment by seungjin-yang** [comment]
Hi @jfernan2, I'm maintaining GEM offline DQM. In [GEM/Efficiency/type1/Efficiency](https://tinyurl.com/yjstnhvf), there are plots about GEM efficiency as measured by global muon tracks. The plot below is the eta distribution of global muon tracks. (I downloaded a `step3.root` file from [here](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-ad6055/17187/runTheMatrix-results/34634.0_TTbar_14TeV+2026D76+TTbar_14TeV_TuneCP5_GenSimHLBeamSpot14+DigiTrigger+RecoGlobal+HARVESTGlobal/).) We can see that there are only six muons and no muons in high eta region (1.5 ~ 3.5) where GEM detectors are installed. So, the empty plots are obviously due to few statistics.

image

**Comment by jfernan2** [comment]
please test

File: DQM/GEM/plugins/GEMRecHitSource.cc [hunk 2]
```diff
@@ -209,7 +209,7 @@ void GEMRecHitSource::analyze(edm::Event const& event, edm::EventSetup const& ev
         Int_t nCLS = hit->clusterSize();
         Int_t nCLSCutOff = std::min(nCLS, nCLSMax_);  // For overflow
         mapCLSRecHit_ieta_.Fill(key3AbsReIEta, nCLSCutOff);
```
Comment (jfernan2) [inline DQM/GEM/plugins/GEMRecHitSource.cc:211]: Is this fill correct too?

Comment (quark2) [reply DQM/GEM/plugins/GEMRecHitSource.cc:211]: Yep, that line has no problem

**Comment by jfernan2** [comment]
please test

**Comment by jfernan2** [comment]
One last question @quark2 @seungjin-yang  do we understand why GEM folder has 480 empty plots? When stats are low, why are these plots created/saved?

https://tinyurl.com/yeqsjmlw

**Comment by quark2** [comment]
image
As Seungjin showed, I also drew the eta distributions of global muons in the [step3 result](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-ad6055/17295/runTheMatrix-results/34634.0_TTbar_14TeV+2026D76+TTbar_14TeV_TuneCP5_GenSimHLBeamSpot14+DigiTrigger+RecoGlobal+HARVESTGlobal/). (which must be same as what Seungjin showed). In the phase2 geometry GEM covers about 1.6 < |η| < 2.4 (except ME0 which covers more high eta region), so that the muons didn't get to GEM. Well, the muons from ttbar (-> WbWb -> 2l 2ν 2b or lν 2q 2b) mostly goes to low eta region, so it seems not strange to me. And actually I don't know whether or not we have not to draw the plots (only in validation...? But these plots are in offlineDQM.) when we have too low stats.

**Comment by jfernan2** [comment]
OK, but would it be possible to mask those efficiency plots if we don't really have muons to compute efficiencies? Thanks

**Comment by quark2** [comment]
I've never heard that kind of masking. Let me ask @seungjin-yang about that. Or could you provide an instruction about that? Thanks!
Btw related to this PR there will be more PRs for offlineDQM and GEM validations. Maybe we can implement the masking in the PRs.

**Comment by jfernan2** [comment]
Thanks @quark2 
By masking I mean that, those plots should only be booked (or at least saved) if you have some efficiency to plot, .i.e. if no muons are registered within GEM acceptance there is no way to compute efficiencies, right?

**Comment by jshlee** [comment]
hi @jfernan2 - I'm not sure if booking only when muons are found is the right solution.
In general, we would like it on for most cases, but for very small unit test like this, it might be better to just turn it off completely.

**Comment by seungjin-yang** [comment]
Hi @jfernan2, `MonitorElement`s for the number of total muons and passed muons are created in `DQM` step. On the other hand, the efficiency plots are created in `HARVESTING` step. So, the harvester can easily control whether to draw efficiency plots or not using the number of total muons. But if we don't plot the efficiency for a small statistic, we don't know how PR affects the efficiency plot because there is no workflow to generate enough muons flying to GEM now. So, in my opinion, masking doesn't seem appropriate unless there is wf suitable for GEM.

**Comment by jfernan2** [comment]
OK, I agree @seungjin-yang

**Comment by jfernan2** [comment]
+1

**Comment by perrotta** [comment]
+1
- It will have to be eventually backported anyhow

**Pull request merged**